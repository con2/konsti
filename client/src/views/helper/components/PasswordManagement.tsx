import React, { FC, ReactElement, ChangeEvent } from 'react';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { getUserBySerial, updateUserPassword } from 'services/userServices';

export const PasswordManagement: FC = (): ReactElement => {
  const { t } = useTranslation();

  const [serial, setSerial] = React.useState<string>('');
  const [username, setUsername] = React.useState<string>('');
  const [passwordFieldType, setPasswordFieldType] = React.useState<string>(
    'password'
  );
  const [userSerialInput, setUserSerialInput] = React.useState<string>('');
  const [changePasswordInput, setChangePasswordInput] = React.useState<string>(
    ''
  );
  const [
    changePasswordInputVisible,
    setChangePasswordInputVisible,
  ] = React.useState<boolean>(false);
  const [userFoundMessage, setUserFoundMessage] = React.useState<string>('');
  const [passwordChangeMessage, setPasswordChangeMessage] = React.useState<
    string
  >('');

  const submitGetUser = async (): Promise<void> => {
    const response = await getUserBySerial(userSerialInput);

    if (!response || response.status === 'error') {
      setUserFoundMessage(`${t('userNotFound')}`);
    } else if (response.status === 'success') {
      setUserFoundMessage(`${t('foundUser')}: ${response.username}`);
      setSerial(response.serial);
      setUsername(response.username);
      setChangePasswordInputVisible(true);
    }
  };

  const submitUpdatePassword = async (): Promise<void> => {
    const response = await updateUserPassword(
      username,
      serial,
      changePasswordInput,
      true
    );
    if (!response || response.status === 'error') {
      setPasswordChangeMessage(`${t('changingPasswordError')}`);
    } else if (response.status === 'success') {
      setPasswordChangeMessage(`${t('changingPasswordSuccess')}`);
    }
  };

  const handleSerialChange = (event: ChangeEvent<HTMLInputElement>): void => {
    setUserSerialInput(event.target.value);
  };

  const handlePasswordChange = (event: ChangeEvent<HTMLInputElement>): void => {
    setChangePasswordInput(event.target.value);
  };

  const togglePasswordVisibility = (): void => {
    if (passwordFieldType === 'password') setPasswordFieldType('text');
    else if (passwordFieldType === 'text') setPasswordFieldType('password');
  };

  return (
    <div className='password-management'>
      <h3>{t('helperPasswordManagement')}</h3>
      <div>
        <p>{t('userCode')}</p>
        <FormInput
          key='user-serial'
          placeholder={t('userCode')}
          value={userSerialInput}
          onChange={handleSerialChange}
        />
        <button onClick={submitGetUser}>{t('button.find')}</button>
        <p>{userFoundMessage && <span>{userFoundMessage}</span>}</p>
      </div>

      <div>
        {changePasswordInputVisible && (
          <>
            <p>{t('newPassword')}</p>
            <FormInput
              type={passwordFieldType}
              key='new-password'
              placeholder={t('newPassword')}
              value={changePasswordInput}
              onChange={handlePasswordChange}
            />
            <button onClick={submitUpdatePassword}>{t('button.save')}</button>

            <FormFieldIcon>
              <FontAwesomeIcon
                icon={passwordFieldType === 'password' ? 'eye' : 'eye-slash'}
                onClick={togglePasswordVisibility}
              />
            </FormFieldIcon>

            <p>
              {passwordChangeMessage && <span>{passwordChangeMessage}</span>}
            </p>
          </>
        )}
      </div>
    </div>
  );
};

const FormFieldIcon = styled.span`
  padding: 0 0 0 8px;
  font-size: ${(props) => props.theme.iconSize};
`;

const FormInput = styled.input`
  border: 1px solid ${(props) => props.theme.borderInactive};
  color: ${(props) => props.theme.buttonText};
  height: 34px;
  padding: 0 0 0 10px;
  width: 100%;
`;
