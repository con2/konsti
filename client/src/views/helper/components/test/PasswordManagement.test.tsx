import React from 'react';
import { shallow } from 'enzyme';
import { PasswordManagement } from '../PasswordManagement';

describe('PasswordManagement', () => {
  it('should render correctly', () => {
    const component = shallow(<PasswordManagement />);
    expect(component).toMatchSnapshot();
  });
});
