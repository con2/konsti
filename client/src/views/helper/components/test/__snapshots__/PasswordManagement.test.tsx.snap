// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordManagement should render correctly 1`] = `
<div
  className="password-management"
>
  <h3>
    helperPasswordManagement
  </h3>
  <div>
    <p>
      userCode
    </p>
    <PasswordManagement__FormInput
      key="user-serial"
      onChange={[Function]}
      placeholder="userCode"
      value=""
    />
    <button
      onClick={[Function]}
    >
      button.find
    </button>
    <p />
  </div>
  <div />
</div>
`;
